src/System.IO/src/System/IO/BufferedStream.cs

```diff
@@ -0,0 +1,1105 @@
+                    _stream.Flush();
+
+                Debug.Assert(_writePos == 0 && _readPos == 0 && _readLen == 0);
+                return;
+            }
+
+            // We had no data in the buffer, but we still need to tell the underlying stream to flush.
+            if (_stream.CanWrite || _stream is BufferedStream)
+                _stream.Flush();
+
+            _writePos = _readPos = _readLen = 0;
+        }
+
+        public override Task FlushAsync(CancellationToken cancellationToken)
+        {
+
+            if (cancellationToken.IsCancellationRequested)
+                return Task.FromCanceled<int>(cancellationToken);
+
+            EnsureNotClosed();
+
+            return FlushAsyncInternal(cancellationToken, this, _stream, _writePos, _readPos, _readLen);
+        }
+
+        private static async Task FlushAsyncInternal(CancellationToken cancellationToken, BufferedStream _this, Stream stream, int writePos, int readPos, int readLen)
+        {
+            // We bring instance fields down as local parameters to this async method becasue BufferedStream is derived from MarshalByRefObject.
+            // Field access would be from the async state machine i.e., not via the this pointer and would require runtime checking to see
+            // if we are talking to a remote object, whcih is currently very slow (Dev11 bug #365921).
+            // Field access from within Asserts is, of course, irrelevant.
+            Debug.Assert(stream != null);
+
+            SemaphoreSlim sem = _this.EnsureAsyncActiveSemaphoreInitialized();
+            await sem.WaitAsync().ConfigureAwait(false);
+            try
+            {
+
+                if (writePos > 0)
+                {
+
+                    await _this.FlushWriteAsync(cancellationToken).ConfigureAwait(false);
+                    Debug.Assert(_this._writePos == 0 && _this._readPos == 0 && _this._readLen == 0);
+                    return;
+                }
+
+                if (readPos < readLen)
+                {
+
+                    // If the underlying stream is not seekable AND we have something in the read buffer, then FlushRead would throw.
+                    // We can either throw away the buffer resulting in date loss (!) or ignore the Flush. (We cannot throw becasue it
+                    // would be a breaking change.) We opt into ignoring the Flush in that situation.
+                    if (!stream.CanSeek)
+                        return;
+
+                    _this.FlushRead();  // not async; it uses Seek, but there's no SeekAsync
+
+                    // User streams may have opted to throw from Flush if CanWrite is false (although the abstract Stream does not do so).
+                    // However, if we do not forward the Flush to the underlying stream, we may have problems when chaining several streams.
+                    // Let us make a best effort attempt:
+                    if (stream.CanRead || stream is BufferedStream)
```

Comment:
This implementation is largely an async-ified version of the sync method, so the same questions I had in the sync method apply here, too, e.g. why this special check for BufferedStream?